sympy/sets/conditionset.py

```diff
@@ -57,67 +58,75 @@ class ConditionSet(Set):
     >>> ConditionSet(n, cond, S.Integers)
     EmptySet
 
-    In addition, substitution of a dummy symbol can only be
-    done with a generic symbol with matching commutativity
-    or else a symbol that has identical assumptions. If the
-    base set contains the dummy symbol it is logically distinct
-    and will be the target of substitution.
+    Only free symbols can be changed by using `subs`:
 
     >>> c = ConditionSet(x, x < 1, {x, z})
     >>> c.subs(x, y)
     ConditionSet(x, x < 1, FiniteSet(y, z))
 
-    A second substitution is needed to change the dummy symbol, too:
-
-    >>> _.subs(x, y)
-    ConditionSet(y, y < 1, FiniteSet(y, z))
+    To check if ``pi`` is in ``c`` use:
 
-    And trying to replace the dummy symbol with anything but a symbol
-    is ignored: the only change possible will be in the base set:
-
-    >>> ConditionSet(y, y < 1, {y, z}).subs(y, 1)
-    ConditionSet(y, y < 1, FiniteSet(z))
-    >>> _.subs(y, 1)
-    ConditionSet(y, y < 1, FiniteSet(z))
+    >>> pi in c
+    False
 
     If no base set is specified, the universal set is implied:
 
     >>> ConditionSet(x, x < 1).base_set
     UniversalSet
 
-    Although expressions other than symbols may be used, this
-    is discouraged and will raise an error if the expression
-    is not found in the condition:
+    Only symbols or symbol-like expressions can be used:
 
     >>> ConditionSet(x + 1, x + 1 < 1, S.Integers)
-    ConditionSet(x + 1, x + 1 < 1, Integers)
-
-    >>> ConditionSet(x + 1, x < 1, S.Integers)
     Traceback (most recent call last):
     ...
     ValueError: non-symbol dummy not recognized in condition
 
-    Although the name is usually respected, it must be replaced if
-    the base set is another ConditionSet and the dummy symbol
-    appears as a free symbol in the base set and the dummy symbol
-    of the base set appears as a free symbol in the condition:
-
-    >>> ConditionSet(x, x < y, ConditionSet(y, x + y < 2, S.Integers))
-    ConditionSet(lambda, (lambda < y) & (lambda + x < 2), Integers)
-
-    The best way to do anything with the dummy symbol is to access
-    it with the sym property.
+    When the base set is a ConditionSet, the symbols will be
+    unified if possible with preference for the outermost symbols:
 
-    >>> _.subs(_.sym, Symbol('_x'))
-    ConditionSet(_x, (_x < y) & (_x + x < 2), Integers)
+    >>> ConditionSet(x, x < y, ConditionSet(z, z + y < 2, S.Integers))
+    ConditionSet(x, (x < y) & (x + y < 2), Integers)
 
     """
-    def __new__(cls, sym, condition, base_set=S.UniversalSet):
```

Comment:
I'm not sure what to do about this: if I make the signature like ImageSet and allow multiple sets to correspond to multiple args then we lose the keyword `base_set` since even if I made the signature `(cls, sym, condition, *base_set)` I couldn't pass the sets by name (e.g. `...base_set=Reals`) only by position (as I understand it).

Reply:
It would be possible by having `*args` and `**kwargs` and manually checking if a keyword argument was supplied.

I'm not sure whether it's a good idea though. I wonder if the multi-arg form of `ImageSet` is a bad idea as it does complicate the definition. It makes some sense for `ImageSet` because it allows to matches a multi-arg `Lambda`. I think it definitely makes less sense for `CondtionSet` though as it would break some symmetries in the n-ary case:
```
ConditionSet(x, p(x), S1, S2, S3)  = {x : p(x), x in prod(S1, S2, S3)} 
ConditionSet(x, p(x), S1, S2)  = {x : p(x), x in prod(S1, S2)}
ConditionSet(x, p(x), S1)  = {x : p(x), x in S1}
```
Here the first two are sets of 3-tuples and 2-tuples. Shouldn't the third be a set of 1-tuples?

We don't need to keep the `ImageSet` in this form though since there is always an alternative. I meant to add a method to `ImageSet` (not sure if I did) that would convert it to single base set form, something like:
```
>>> ImageSet(Lambda((x, y, z), x+y+z, S1, S2, S3).canonical()
ImageSet(Lambda(((x, y, z),), x+y+z, ProductSet(S1, S2, S3))
```
Perhaps in code that wants to relate `ImageSet` and `ConditionSet` we should just always call the canonical method. Maybe canonical isn't a good name and it could be flatten (or unflatten?) or a property rather than a method. Not sure... Maybe ImageSet should automatically canonicalise itself... or maybe doit should?

Reply:
> ConditionSet(x, p(x), S1, S2, S3)  = {x : p(x), x in prod(S1, S2, S3)}

The way I have it set now, this would raise an error since `x` is passed; it would not raise an error for `ConditionSet((x,y,z), p(x,y,z), S!, S2, S3)`. I don't think we would want to allow the form you gave to pass or else how would the individual components of x be used in the predicate?

Reply:
Does that not leave an ambiguity:
```
ConditionSet((x, y, z), p(x, y, z), S)
ConditionSet((x, y, z), p(x, y, z), S1, S2, S3)
```
Then in the first case `(x, y, z)` is an element of `S`. In the second case it is an element of `prod(S1, S2, S3)`. Then in the n-ary case we should have
```
syms = tuple([Dummy() for _ in range(n))
ConditionSet(syms, 
```

Reply:
ConditionSet canonicalizes the base set to the `...S1, S2, S3` is interpreted (and stored) as `ProductSet(S1, S2, S3)`.

Reply:
In the first case, the base set is interpreted as the base set for the properly dimensioned `base_set`; it is shorthand for saying "all individual syms are of the same type"

Reply:
I edited my last comment (sent too early)

Reply:
I think this is how I ended up putting a single symbol in a tuple in Lambda:
```
In [2]: Lambda(x, x).signature                                                                                                                 
Out[2]: (x,)

In [3]: Lambda((x,), x).signature                                                                                                              
Out[3]: (x,)

In [4]: Lambda((x, y), x).signature                                                                                                            
Out[4]: (x, y)
```

Reply:
Is the difference between S and ProducSet(S) like the difference between x and (x,)?

> ConditionSet(x, p, X)
> ConditionSet((x,), p, ProductSet(X))
> ConditionSet((x, y), p, ProductSet(X, Y))
> ConditionSet((x, y, z), p, ProductSet(X, Y, Z))

I can't test this now...I suspect that what I have will allow this and `ConditionSet((x,),p,S)` will correctly become `C((x,),p,ProductSet(S))` with the morphbase handling.

Reply:
I added a test for that. A tuple-sym will have a ProductSet base_set after instantiation by merit of 1) it being produced to match the structure of sym and all sets being the same, e.g. `(Integers, (Integers, Integers))` for `C((x,(y,z)),p,Integers)` or 2) a check that dimensions of sym match the dimensions given as either a proper ProductSet for the `base_set` arg or ProductSet args as in `C((x, (y, z)), p, S1, S2*S3)`. The only reason I am allowing this syntax is to copy ImageSet acceptance of the same.

If there are no other issues I think this is ready to go.

Reply:
> Is the difference between S and ProducSet(S) like the difference between x and (x,)?

Yes

Reply:
The morphbase handling seems quite complicated

Reply:
I think it would be good to try and start with a clean definition that doesn't require special cases. The idea with Lambda/ProductSet is just that a single symbol is considered an acceptable "short notation" for a 1-tuple. After translating that to a 1-tuple the rules are that the length of the tuple matches the number of arguments/base sets. Any nested tuples in the signature correspond to sets of tuples in the base set which are usually but *not always* represented as ProductSets.

Reply:
> Any nested tuples in the signature correspond to sets of tuples in the base set which are usually but _not always_ represented as ProductSets.

Can you give an example of what you mean (because *that* sounds complicated to me).

Reply:
> The morphbase handling seems quite complicated

It recursively creates `ProductSet(*([base_set]*len(t)))` for tuples of symbols. Do you just not like being able to apply a single Set to all symbols in `sym`?

Reply:
> > Is the difference between S and ProducSet(S) like the difference between x and (x,)?
> 
> Yes

OK, then, I have confirmed that it is handled properly in each case.

Reply:
Coherent signature/base_set is not enforced in master 
```python
>>> ConditionSet(x,x>1,Integers)
ConditionSet(x, x > 1, Integers)
>>> ConditionSet((x,),x>1,Integers)
ConditionSet((x,), x > 1, Integers)
>>> ConditionSet((x,y),x>1,Integers)
ConditionSet((x, y), x > 1, Integers)
>>> (2,3) in _
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "\sympy\sets\sets.py", line 669, in __contains__
    c = self._contains(other)
  File "\sympy\sets\conditionset.py", line 204, in _contains
    Lambda(self.sym, self.condition)(other))
  File "\sympy\core\function.py", line 2039, in __call__
    'given': n})
sympy.core.function.BadArgumentsError: Lambda((x, y), x > 1) takes exactly 2 arguments (1 given)
```
so I guess we can make it be whatever we want here. Would you be ok with
```
C(x, p, S)
C((x,y), p, S) -> C((x,y), p, S**2)
C((x,y), p, S1*S2) but not C((x,y), p, S1, S2)
```
The 2nd case is incompatible with what ImageSet does
```python
>>> ImageSet(Lambda((x,(y,z)),x+y/Max(z+1,1)),Integers*Integers**2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "\sympy\sets\fancysets.py", line 336, in __new__
    raise ValueError('Incompatible signature')
ValueError: Incompatible signature
```
(an accepted base set for this case is `(S, S**2)`)
```
> usually but _not always_ represented as ProductSets.

but they have to be in order to evaluate the imageset...otherwise you will get an error:

```
>>> i3=ImageSet(Lambda((x,(y,z)),x+y/Max(z+1,1)),Integers, Integers**2)
>>> it=iter(i3)
>>> [next(it) for i in range(15)]
[0, 1, -1, 2, -2, 3, 1/2, 3/2, -3, 4, -1/2, 5/2, -4, 5, -3/2]
>>> ImageSet(Lambda((x,(y,z)),x+y/Max(z+1,1)),Integers, Integers)
ImageSet(Lambda((x, (y, z)), x + y/Max(1, z + 1)), Integers, Integers)
>>> it=iter(_)
>>> [next(it) for i in range(15)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <listcomp>
  File "\sympy\sets\fancysets.py", line 397, in __iter__
    val = self.lamda(*i)
  File "\sympy\core\function.py", line 2041, in __call__
    d = self._match_signature(self.signature, args)
  File "\sympy\core\function.py", line 2058, in _match_signature
    rmatch(sig, args)
  File "\sympy\core\function.py", line 2055, in rmatch
    raise BadArgumentsError("Can't match %s and %s" % (args, pars))
sympy.core.function.BadArgumentsError: Can't match (0, 0) and (x, (y, z))
>>>
```

Reply:
> The 2nd case is incompatible with what ImageSet does
> 
> ```python
> >>> ImageSet(Lambda((x,(y,z)),x+y/Max(z+1,1)),Integers*Integers**2)
> ```

The signature should be wrapped in an extra tuple:
```
In [2]: ImageSet(Lambda(((x,(y,z)),),x+y/Max(z+1,1)),Integers*Integers**2)                                                                     
Out[2]: 
⎧          y                           ⎛ 2⎞⎫
⎨x + ───────────── | (x, (y, z)) ∊ ℤ × ⎝ℤ ⎠⎬
⎩    Max(1, z + 1)                         ⎭
```
This is because the number of arguments of the `Lambda` needs to match the number of basesets given as args to `ImageSet`.

The idea with the way ImageSet is arranged is that we have the simple definition:
```
IS = ImageSet(f, S1, S2, ...)

IS is the set such that:
If x1 in S1, x2 in S2, ... and args = (x1, x2, ...) then f(*args) is in S1
```
Note that this definition does not require introspecting the the signature of f or looking at the types of sets S1, ...

We can't in general know if a set is a set of tuples or not and it is allowed that a symbol in the signature might represent a tuple:
```
In [3]: F = FiniteSet((1, (2, 3)), (4, (5, 6)))                                                                                                

In [4]: ImageSet(Lambda(((x, y),), y), F)                                                                                                      
Out[4]: {y | (x, y) ∊ {(1, (2, 3)), (4, (5, 6))}}

In [5]: ImageSet(Lambda(((x, y),), y), F).doit()                                                                                               
Out[5]: {(2, 3), (5, 6)}
```

The only time we can have a bad signature is when the signature is a tuple of known length and the set is a productset of the wrong length:
```
In [9]: ImageSet(Lambda(((x, y),), y), Reals**3)                                                                                               
---------------------------------------------------------------------------
ValueError
```

Reply:
> We can't in general know if a set is a set of tuples or not

Can you give an example? We have a finite number of types of sets and each type of set has args and those args must be able to match up with the signature so I can't think of why this should be anything other than tedious, not impossible.

Reply:
> This is because the number of arguments of the `Lambda` needs to match the number of basesets given as args to `ImageSet`.
> 
> The idea with the way ImageSet is arranged is that we have the simple definition:

This was a helpful reminder. Should ConditionSet should behave in the same way?

Reply:
> Can you give an example?

A FiniteSet can have anything including tuples and non-tuples and tuples of mixed length `FiniteSet((x, y), (x, y, z), 1)`. A ConditionSet, ImageSet or ProductSet could be based on a FiniteSet. Same goes for Intersection etc.

> Should ConditionSet should behave in the same way?

Maybe. Perhaps we could define a `ConditionSet` as
```
CS = ConditionSet(sig, cond, S1, S2, ...)
f = Lambda(sig, cond)
If x1 in S1, x2 in S2, ... and elem = (x1, x2, ...) then elem is in CS if `f(*elem)`
```
That definition does mean that when you have a single base set the set is a set of 1-tuples though. This is a fundamental problem really because it isn't clear what the ConditionSet is a subset of any more. In the multi base set case it's a subset of `ProductSet(S1, S2, ...)` but that can't carry over to the single base set case.

Really I think that the multiple base set form of `ImageSet` is a mistake. We already have `ProductSet` without needing to incorporate `ProductSet` functionality into `ImageSet` and `ConditionSet`. I've extended the functionality of `Lambda` so that you can do anything you want with a multivariate `ImageSet` over a single `ProductSet` base set.

If we don't have the multi arg form then the definitions of both are very simple and clear:
```
ImageSet(f, S) == {f(x): x in S}
ConditionSet(sig, cond, S)  == {x : x in S and g(x)}   (where g = Lambda(sig, cond))
```

Reply:
Should I close this PR?

Reply:
> A FiniteSet can have anything including tuples and non-tuples and tuples of mixed length

I suppose there could be some sort of predicate that would not require uniformly shaped elements in a FiniteSet...that seems a little obtuse to me, however.

Reply:
There are lots of good changes here but I think it would be better to stick to only having a single base set and not modify the base set to make a ProductSet in `ConditionSet((x, y), x<0, Integers)`.

> some sort of predicate that would not require uniformly shaped elements

You could have something like:
```julia
In [1]: ConditionSet((x, y), x<0, {(1, (2, 3)), (4, (5, 6, 7))})                                                                               
Out[1]: {(x, y) | (x, y) ∊ {(1, (2, 3)), (4, (5, 6, 7))} ∧ (x < 0)}
```
In this PR that comes out strangely:
```julia
In [2]: ConditionSet((x, y), x<0, {(1, (2, 3)), (4, (5, 6, 7))})                                                                               
Out[2]: 
⎧                                               2          ⎫
⎨(x, y) | (x, y) ∊ {(1, (2, 3)), (4, (5, 6, 7))}  ∧ (x < 0)⎬
⎩                                                          ⎭
```

Reply:
> not modify the base set to make a ProductSet


>You could have something like:
> In [1]: ConditionSet((x, y),

So should we just let errors rise when the user tries to do anything with the set?

Reply:
> So should we just let errors rise when the user tries to do anything with the set?

I think for now it's fine to do that. The most important thing to pin down is what exactly is the definition of ConditionSet when given valid arguments.

If we wanted to handle the invalid arguments better at the outset we could add some kind of `is_tuple_set` property to sets that tells you whether it could possibly be a tuple in general (It could be always True for ProductSet, always None for FiniteSet, False for most atomic sets, `fuzzy_and` for `FiniteSet`, `fuzzy_or`? for `Intersection`). We know e.g. that Integers does not contain any tuples. Then a case like `ConditionSet((x, y), x<0, Integers)` could raise by checking `baseset.is_tuple_set is False`.